dill/settings.py

```diff
@@ -19,7 +21,10 @@
     'fmode' : 0, #HANDLE_FMODE
     'recurse' : False,
     'ignore' : False,
+    'dump_module' : {
+        'refimported': False,
+        'refonfail' : False,
```

Comment:
Should the default be `True`?  I'm not suggesting it should be, but just want us to discuss it a bit.

Reply:
Upsides:
- Less error messages in general
- Would lower the steepness of the learning curve and work out-of-the-box for most cases
- More people using this feature may unveil edge cases and let to a higher coverage of modules in the wild

Downsides:
- Would disable framing (speed optimization) by default, although I have no idea of the real impact 
- The particular state the user is interested in saving may not be saved (and they won't notice unless they activate and follow the tracing log or until they verify it worked after restoring the module)
- Currently, the whole module may end being saved by reference, i.e. no state saved at all...

(I'm thinking if this last one shouldn't raise a warning at least.)

I'm divided, but I think it's OK to have this option on by default.

Reply:
To me, it feels like the default should be `True`.  I expect that most people are going be interested in pickling unmodified modules, and "just want it to work".  So, if the majority of people will not use the default, then it's likely the wrong default. I'm fine with:

> the whole module may end being saved by reference

because, that is what `pickle` would give you.  I think it's also going to be expected that `dill` should do at least what `pickle` does, and using a default of `False` may give an impression that this is not the case.

If changing the default to `True`, we'd need to:

1. Document that `refonfail=False` may lead to faster serialization, but may fail when something in the module fails to serialize. 
2. Document that modifications will be ignored in any module that has to be stored by reference.

I think this is pretty acceptable, and a better choice.

Reply:
Everything but the item 2 was applied. But I'm not sure of what to do about objects in the module's namespace (not child objects) that fail to serialize. Consider these scenarios with two imported modules, `mod`  and `other_mod`, and an unpickleable variable named `var` that doesn't have an attribute `__name__` or `__qualname__` (note that there's no way to know where `var` was defined):

1.
    - `mod` defines `var`
    - `var` is not imported into other modules or is imported with aliases
    -  in `dump_module(module=mod)`, `var` may saved globally as `mod.var` or fail to save (e.g. for `__main__`)
2.
    - `mod` defines `var`
    - `other_mod` imports `var` from `mod`
    - in `dump_module(module=mod)`, `var` may saved globally as `mod.var` or `other_mod.var`
3.
    - `other_mod` defines `var`
    - `mod` imports `var` from `other_mod`
    - in `dump_module(module=mod)`, `var` may saved globally as `mod.var` or `other_mod.var`
4.
    - `other_mod` defines `var`
    - `mod` imports `var` as `var_alias`
    - in `dump_module(module=mod)`, `var` may saved globally as `mod.var_alias` or fail to save

**For each of these cases, which result is more appropriate?** Are all the desired behaviors compatible, i.e. is it possible to implement all of them consistently? Remember that, for the third scenario, `other_mod` may be a built-in module, which is safe, but may also be a third-party module not available in the unpickling environment. The behavior in this case can be stratified according to whether `other_mod` is a built-in module, is another module in the same package as `mod`, or is a third-party module.

Reply:
My previous comment brought up some situations where there's an unpickleable object in the namespace of the module being saved by `dump_module()`. If it has `__module__` and `__(qual)name__` attributes (5th scenario), it may be saved as global normally.  If that fails, then it falls into the same situation of an object without these attributes (1st to 4th scenarios).  I need to highlight that in these cases, the situations (2) and (3) are indistinguishable, there's no way to know where the object was originally defined.

Here, I'll further stratify the scenarios (1), (2) plus (3), and (4) presented earlier in a more practical way.  For the following discussion, a "builtin" module is a module for which `_is_builtin_module(module)` returns `True`, and a "user" module is a module for which the same test returns `False`.

Proposition: Assume an unpickleable object named `name` in the namespace of module `mod`. 

Possible cases:

1. `name` is only defined in `mod`
2. `name` is also defined in the "builtin" module `other`
3. object is defined in the "builtin" module `other` as `altname`
4. `name` is also defined in the "user" module `other` (except `__main__`)
5. object is defined in the "user" module `other` (except `__main__`) as `altname`

<table><thead>
  <tr>
    <th>Settings</th>
    <th colspan=2><code>refonfail=True</code> (default)</th>
    <th colspan=2><code>refimported=True</code></th>
    <th colspan=2>Both are <code>True</code></th>
    <th colspan=2>Both are <code>False</code></th>
  </tr>
  <tr>
    <th>Type of <code>mod</code></th>
    <th>"builtin"</th><th>"user"</th>
    <th>"builtin"</th><th>"user"</th>
    <th>"builtin"</th><th>"user"</th>
    <th>"builtin"</th><th>"user"</th>
  </tr>
</thead><tbody>
  <tr align="center">
    <td>Case (1)</td>
    <td>"mod.name"</td><td>exception</td>
    <td colspan=2>exception</td>
    <td>"mod.name"</td><td>exception</td>
    <td colspan=2>exception</td>
  </tr>
  <tr align="center">
    <td>Case (2)</td>
    <td colspan=2>"other.name"</td>
    <td colspan=2>"other.name"</td>
    <td colspan=2>"other.name"</td>
    <td colspan=2>exception</td>
  </tr>
  <tr align="center">
    <td>Case (3)</td>
    <td colspan=2>"other.altname"</td>
    <td colspan=2>"other.altname"</td>
    <td colspan=2>"other.altname"</td>
    <td colspan=2>exception</td>
  </tr>
  <tr align="center">
    <td>Case (4)</td>
    <td>"mod.name"</td><td>"other.name"</td>
    <td colspan=2>"other.name"</td>
    <td colspan=2>"other.name"</td>
    <td colspan=2>exception</td>
  </tr>
  <tr align="center">
    <td>Case (5)</td>
    <td>"mod.name"</td><td>"other.altname"</td>
    <td colspan=2>"other.altname"</td>
    <td colspan=2>"other.altname"</td>
    <td colspan=2>exception</td>
  </tr>
</tbody></table>

Rationale:

Focusing on the first two columns, which define the behavior of `refonfail`, there are two main cases: the module is a "builtin" module (e.g. `math` or installed `numpy`) or a "user" module (e.g. `__main__` or a module in the working directory).  And we have two main considerations to balance: how often should the operation fail (mainly concerning the default case) and wheter it's safe to consider that the object saved as global (without knowing if it actually belongs to the other module) will be there at unpickling.

### "builtin" modules

For "builtin" modules being saved, with the default options `refonfail=True` and `refimported=False`, `dump_module()` should work out-of-the-box, without errors.  As unpickleable objects in these modules are likely to always be defined, it's fairly safe to save them as references to the module itself (case 1, builtin).  Also if there's an object that is only found elsewhere in "user" modules, it should be saved in the same way (cases 4 and 5, builtin), as the most likely scenario is that such object was defined in the "builtin" module being saved and imported into the "user" modules (which have smaller chances to be available in the unpickling environment). Last, if the object is defined in another "builtin" module, it can be saved as a reference to that (cases 2 and 3, builtin).

### "user" modules

For "user" modules, which include `__main__`, if the unpickleable object is defined only in the module being saved, it's safer to assume the object may not be there when unpickling and therefore to raise an error (case 1, user).  While if the object is only defined in other "user" modules (except `__main__`), it may be fair to save it as a reference to one of them (cases 4 and 5, builtin), or alternatively to raise an error too.  This last case is the single one I'm not sure of how it should behave. Cases 2 and 3 should do the same as for "builtin" modules.

### Save the module by reference?

For "builtin" modules, unpickleable objects will always be saved by reference somehow (to a name in another module or to the module itself) with `refonfail=True`.  With `refonfail=False`, we can either save the module by reference, preferebly with a warning, or raise an error.  As this is not the default, I think that raising an error is fair, suggesting to the user the use of filters to exclude the offending object from the dump.

For "user" modules with default options, an unpickleable object will only raise an exception if it can't be found elsewhere (or only in `__main__`).  This case could force the module to be saved by reference and raise a warning.  For `refonfail=False`, I think it should always raise an error.

### Summary

The general behavior can be described as:

- For "builtin" modules, unpickleable objects are always saved by refence, even if it is a reference to the module itself.
- For "user" modules, it may raise an exception (internally) in some cases.
 
When a exception is raised:

- If `refonfail=True`, the default, save the module itself by reference with a warning.
- If `refonfail=False`, raise an error, so that the user has the opportunity to use exclusion filters to be able to save the state of other objects in the module.

I think it's no so hard to implement this logic as we already have `_is_builtin_module()`, `_locate_function()`, `_module_map()` and `_lookup_module()`. I'll try this tomorrow.

What do you think?

Reply:
I think your proposal makes a lot of sense, and I have questions about the same cases as you have questions: (1) what to do for `refonfail=True` case 4/5 for "user", and (2) error vs warning behavior in some cases.

> While if the object is only defined in other "user" modules (except __main__), it may be fair to save it as a reference to one of them (cases 4 and 5, builtin), or alternatively to raise an error too. This last case is the single one I'm not sure of how it should behave.

I'd say, since it's `refonfail=True`... then this makes sense.  I'm wondering if there's a case where it fails in `mod` and is then saved by reference to `other.name` or similar... but the object is actually defined in `mod`, and `other.name` is just a reference to the original in `mod`.  The question is: is it possible for this to go wrong... where save-by-reference to `other.name` is the wrong thing to do.  I think we need to try to think up failure scenarios for this case.

> With refonfail=False, we can either save the module by reference, preferebly with a warning, or raise an error. As this is not the default, I think that raising an error is fair, suggesting to the user the use of filters to exclude the offending object from the dump.

I think I agree here.

> For "user" modules with default options, an unpickleable object will only raise an exception if it can't be found elsewhere (or only in __main__). This case could force the module to be saved by reference and raise a warning.

See my first comment.

> For refonfail=False, I think it should always raise an error.

I agree here.

Generally, I think a very solid and logical proposal.  I may need to construct some test cases and explore how they fail or succeed... this should better inform the cases we are unsure about.

Reply:
Good that you agreed with most of my reasoning. I'll answer to your considerations later today.

While I'm advancing with the draft implementation of this logic, I'd like to propose a hierarchy of "safeness" for modules to be referred to when saving by reference. If an object is found in multiple modules, the safest alternative would always be preferred, as it has more chances to be there at unpickling (it also applies to `refimported`).

The proposed hierarchy is:
1. Standard Library modules
2. modules of the same package as the module being saved
3. installed modules in general ("builtin" modules)
4. non-installed modules ("user" modules)
5. `__main__` — never safe

Cases 1 to 3 would fall under the "builtin" umbrella of the table from my previous comment.

Reply:
I'm not sure what you mean by "safeness" or "safe".  Is it just more granularity to "user" vs "builtin"?

Reply:
"safer" == "greater probability of the module being available in the unpickling environment and of the object being defined in it (and pertain to it)"

For example, consider a module named `pack.mod`. At loading its state with `load_module()`, we can assume that both Standard Library modules and other modules from the package `pack` are available (otherwise, `pack.mod` itself wouldn't be loadable). So, if at saving its state with `dump_module()`, the unpickleable object `pack.mod.obj` can be saved globally either as `std.lib.obj` or as `pack.other.obj`, should we prefer one over the other? I argue that yes, because the most likely scenario is that the object is defined in the Standard Library module `std.lib` and imported from it in both `pack.mod` and `pack.other`. Plus, there's a small chance that, in the unpickling environment, the module `pack.mod` may exist but `pack.other` may not (for example, if we had slightly different versions of the package). For StdLib packages, however, that's highly unlikely.

That's the logic behind the whole hierarchy. It's an empirical way of maximizing the probability of success of saving by reference in the module's namespace.

This is almost orthogonal to the stratification of self/"builtin"/"user" behavior from that table. The only change is that modules that are part of the same package as the module being saved (if the module is part of any package) are put together with the "builtin" modules even if the package is not appropriately installed. Basically, this hierarchy will define which reference to another module is better, but not if that should be used instead of a reference to the module itself, or if an exception should be raised.

Reply:
(I likely wrote this a week ago but didn't press "send"...)

"safer" == "greater probability of the module being available in the unpickling environment and of the object being defined in it (and pertain to it)"

For example, consider a module named `pack.mod`. At loading its state with `load_module()`, we can assume that both Standard Library modules and other modules from the package `pack` are available (otherwise, `pack.mod` itself wouldn't be loadable). So, if at saving its state with `dump_module()`, the unpickleable object `pack.mod.obj` can be saved globally either as `std.lib.obj` or as `pack.other.obj`, should we prefer one over the other? I argue that yes, because the most likely scenario is that the object is defined in the Standard Library module `std.lib` and imported from it in both `pack.mod` and `pack.other`. Plus, there's a small chance that, in the unpickling environment, the module `pack.mod` may exist but `pack.other` may not (for example, if we had slightly different versions of the package). For StdLib packages, however, that's highly unlikely.

That's the logic behind the whole hierarchy. It's an empirical way of maximizing the probability of success of saving by reference in the module's namespace.

This is almost orthogonal to the stratification of self/"builtin"/"user" behavior from that table. The only change is that modules that are part of the same package as the module being saved (if the module is part of any package) are put together with the "builtin" modules even if the package is not appropriately installed. Basically, this hierarchy will define which reference to another module is better, but not if that should be used instead of a reference to the module itself, or if an exception should be raised.

Reply:
Answering your previous comment:

> > While if the object is only defined in other "user" modules (except **main**), it may be fair to save it as a reference to one of them (cases 4 and 5, builtin), or alternatively to raise an error too. This last case is the single one I'm not sure of how it should behave.
> 
> I'd say, since it's `refonfail=True`... then this makes sense. I'm wondering if there's a case where it fails in `mod` and is then saved by reference to `other.name` or similar... but the object is actually defined in `mod`, and `other.name` is just a reference to the original in `mod`. The question is: is it possible for this to go wrong... where save-by-reference to `other.name` is the wrong thing to do. I think we need to try to think up failure scenarios for this case.

This problem of saving a reference to an object of other module that we don't know where was defined is the same that already existed for the `refimported` option (the old `byref` for modules). It could go wrong in many cases, yes, although I think they are a small minority of cases.

What I could think of to minimize this was: get the module associated with the object. It may be `obj.__module__` or `type(obj).__module__`. If it matches the module being pickled, don't save it as a reference to other modules, as it's most likely that the object was defined in the module itself. Secondly, if the module being pickled is part of a package, and the module associated with the object is from this package, only save it as a reference to a module of the same package.

The reason for the second rule is that packages in general are independent and have a unidirectional relationship, i.e. one acts as a library and the other as a client. So, suppose we are pickling a module `pack1.bar`, and we hit an unpickleable object of type `pack1.baz.Class`. We then find the same object also defined in a module `pack2.foo`. The most probable scenario is that `pack2` uses `pack1` as a library and `pack2.foo` imported the object from `pack1.bar` (the module we are pickling), not the opposite. The alternative hypothesis is that `pack2.foo` used `pack1.baz.Class` to create the object and then injected it in the `pack1.bar` namespace, which is very unlikely.

> Generally, I think a very solid and logical proposal. I may need to construct some test cases and explore how they fail or succeed... this should better inform the cases we are unsure about.

Well, I implemented tests to check the stratified behavior and the hierarchy of preferences when there are multiple choices of references to use.

Do you want some tests that would fail? I can think of a couple ones.

Reply:
When a package has some fraction of a billion users, the thing you have to remember is that any time you think it's a rare corner case, someone will be attempting that rare corner case.  So, then the assessment is: what is the cost if `dill` serializes something incorrectly? Or what happens later to users with stored pickle files if the behavior is then corrected?  An at first seemingly rare, but actually somewhat common, is your latter example but the package is so large it has sub-packages (so like `numpy`, `sklearn`, `scipy`, etc), so you'd have `pack.pack2.foo` using `pack.pack1.baz.Class` to create an object and inject it into the `pack.pack1.bar`... which is suddenly a lot less unlikely.

Reply:
Sorry, I wasn't clear. I did notice that while implementing these package-related logic. With "package" I meant "top-level package", not the subpackage (so `module.__package__.partition('.')[0]` when there's a package value) . I tested it on all modules loaded by default by `numpy`, `pandas`, `scipy`, `sklearn` and also `IPython`, and this value was consistent.

Look, if you are not 100% confortable with this feature, maybe it's better to leave it off by default in this release, see how it's used in the wild and if bugs/problems come up and then consider enabling it by default in a future release if we feel more confident.

A different thing that I thought of was to have a more granular control of for which modules these settings (`refonfail` and `refimported`) are enabled. We could have more options, something like (thinking of a bit field):
- 0 or False: disabled
- 1 or True: enabled for all modules
- 2: enabled for Standard Library modules
- 4: enabled for installed packages
- 8: enabled for user modules
- 16: enabled for __main__
- 32: enabled for module-type objects

Usage would be similar to the `os` flags for `open()`, like:
```python
import dill
from dill.session import settings
settings['refonfail'] = dill.MOD_STDLIB | dill.MOD_INSTALLED
```

Reply:
> Sorry, I wasn't clear. I did notice that while implementing these package-related logic. With "package" I meant "top-level package", not the subpackage (so `module.__package__.partition('.')[0]` when there's a package value). I tested it on all modules loaded by default by `numpy`, `pandas`, `scipy`, `sklearn` and also `IPython`, and this value was consistent.

Excellent.  Thanks for the clarification.  Also, your idea about granular control is interesting. I don't think it should be addressed right now, however.  Post-release.

Reply:
Right, I agree. I would only add this granular control now to have as default the `refonfail` option enabled only for Standard Library modules (and maybe for installed user modules). If we'll keep the current default, then leave that for later.